core/trino-main/src/main/java/io/trino/util/Executors.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.util;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletionService;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorCompletionService;
+import java.util.concurrent.Future;
+
+import static io.airlift.concurrent.MoreFutures.getDone;
+
+public final class Executors
+{
+    private Executors() {}
+
+    public static <T> void executeUntilFailure(Executor executor, Collection<Callable<T>> tasks)
+    {
+        CompletionService<T> completionService = new ExecutorCompletionService<>(executor);
+        List<Future<T>> futures = new ArrayList<>(tasks.size());
+        for (Callable<T> task : tasks) {
+            futures.add(completionService.submit(task));
+        }
+        try {
+            for (int i = 0; i < futures.size(); i++) {
+                getDone(take(completionService));
+            }
+        }
+        catch (Exception failure) {
+            try {
+                futures.forEach(future -> future.cancel(true));
+            }
+            catch (RuntimeException e) {
+                failure.addSuppressed(e);
+            }
```

Comment:
should we wait for other futures to complete?

Reply:
I decided to fail fast.